### ISSUE TITLE
Undefined props not handled correctly.

### ISSUE THREAD

**Issue by arackaf** [preactjs/preact#1943]
Per Jason, this line

https://github.com/preactjs/preact/blob/master/src/diff/props.js#L17 

causes incorrect behavior for hyperlinks in some rare cases.

**Comment by JoviDeCroock** [comment]
Do you mean when you do `<a href="/" />` and next render `<a />`? Does this produce `<a href="" />` or? Sorry having trouble following

**Comment by developit** [comment]
Apologies, I'd meant to add some context but forgot.
Currently, the following will leave an `href` attribute in-place in the DOM, because the comparison on that line considers missing props and undefined props to be the same:

```js
function Demo() {
  const [props, setProps] = useState({ href: undefined });
  useEffect(() => {
    setProps({ });  // no href
  }, []);
  return <a {...href}>demo</a>
}
render(<Demo/>, document.body);
```

Here's a reproduction I set up that illustrates:
https://jsfiddle.net/developit/jn60arh5/

**Comment by ChrisAckerman** [comment]
I encountered a related issue to this. A third party component was using `Object.assign({ ref: () => {...} }, props)`, expecting to inject a `ref` callback if no default ref callback was provided. But, due to the above (probably?) the incoming props **always** has its own `ref` property, it's just set to `undefined` most of the time. And, Object.assign faithfully overwrites the fallback `ref` with `undefined` because it copies own enumerables, even if they're undefined.

I fixed it as follows (typescript):

```tsx
import { options } from "preact";

const _vnode = options.vnode;

options.vnode = (vnode) => {
  if (vnode.props && vnode.props.hasOwnProperty("ref")) {
    delete (vnode.props as any).ref;
  }
  if (_vnode) _vnode(vnode);
};
```
